### DISCUSSION TITLE
Questions about Styled/Direct properties and Routed events

### DISCUSSION THREAD

**Discussion by HavenDV** [AvaloniaUI/Avalonia#8543] [Q&A] [answered]
I'm adding support for Avalonia in [DependencyPropertyGenerator](https://github.com/HavenDV/DependencyPropertyGenerator) (all examples from the documentation are covered at the moment, including AddOwner) and would like to ask a series of questions about the appropriateness of different combinations of properties:
1. Do attached direct properties make sense?
2. Does AddOwner make sense for an attached property?
3. Do attached routed events make sense and what is the pattern of their declaration?
4. What does `notifying` mean for regular Styled properties? When it can be used, is it equivalent to the Changing callback for other frameworks?
5. Does read-only attached StyledProperty make sense?

**Comment by maxkatz6** [comment, marked as answer]
Hi! Thanks for your interesest in adding Avalonia support. I will try to answer what I can, and maybe @grokys can take a look.

> Do attached direct properties make sense?

No. Direct properties are backed with a field, and you can't "attach" a new field to another class.

> Does AddOwner make sense for an attached property?

If you want to create CLR property in your control which will reuse existing attached property, then it's better to use AddOwner. But from what I remember it's not necessary, it would work without it.

 > Do attached routed events make sense and what is the pattern of their declaration

Technically routed events are not bound to the specific control. Which means you can call `textBox.RaiseEvent(new RoutedEventArgs(Button.ClickEvent))`. I suppose it would work as an attached event in this case, with no difference in the event declaration.

> What does notifying mean for regular Styled properties ? When it can be used

Not sure about this one. From what I remember this one has more use for Avalonia internals, and wasn't intended to be used by developers. But it's possible though. 

> is it equivalent to the Changing callback for other frameworks?

There are two recommended ways to handle changes:
1. `override void OnPropertyChanged` method. This should have the lowest overhead. You can filter it by the property in the parameters. But if you use this one in the source generated code, developers might have a problem if they want to override it by themselves in the same class.
2. Reactive subscription `Popup.IsOpenProperty.Changed.Subscribe`.

> Does read-only attached StyledProperty make sense?

No, it doesn't (in context of Avalonia at least).
Only readonly direct property make sense. I.e. direct property without any defined setter.

**Reply by HavenDV** [reply]
Thanks for the quick response!

> 2. Reactive subscription `Popup.IsOpenProperty.Changed.Subscribe`.

Here I meant the Changing event, not Changed. MAUI has this capability in conjunction with the Changed event. Therefore, I was clarifying if the Notifying event is similar to this.

> Technically routed events are not bound to the specific control. Which means you can call textBox.RaiseEvent(new RoutedEventArgs(Button.ClickEvent)). I suppose it would work as an attached event in this case, with no difference in the event declaration.

I thought there might be some pattern here to define static methods to use this in axaml files, like SetValue/GetValue static methods for attached properties.
